## Unión y depuración de la data

En este notebook uniremos toda la data recopilada anteriormente y crearemos un único DataFrame.  
Posteriormente realizaremos una limpieza de los datos.

In [7]:
import glob
import pandas as pd
from functools import reduce
import numpy as np
pd.options.mode.chained_assignment = None #evitar la advertencia: SettingWithCopyWarning

En lo posterior, necesito que el notebook muestre todas las columnas del DataFrame:

In [8]:
pd.set_option('display.max_columns', None)

Unimos todos los csv para crear un único DataFrame:

In [9]:
dfLista = []
ruta = '/home/dsc/Python/Despoblacion/Data' 
csv_files = glob.glob(ruta + "/*11.csv")
csv_files.sort()

In [10]:
for file in csv_files:
    df = pd.read_csv(file, dtype = {'Codigo Provincia' : str, 'Codigo Municipio' : str})
    dfLista.append(df)

In [11]:
Despoblacion = reduce(lambda  left,right: pd.merge(left,right,on=['Codigo Municipio'],
                                            how='left'), dfLista)
Despoblacion.head()

Nombre Provincia Codigo Provincia  Nombre Municipio Codigo Municipio  \
0      Araba/Álava               01  Alegría-Dulantzi            01001   
1      Araba/Álava               01           Amurrio            01002   
2      Araba/Álava               01             Añana            01049   
3      Araba/Álava               01           Aramaio            01003   
4      Araba/Álava               01           Armiñón            01006   

   Poblacion Total  Poblacion Hombre  Poblacion Mujer  Superficie Km2  \
0           2803.0            1466.0           1337.0           19.95   
1          10014.0            5030.0           4984.0           96.30   
2            168.0              86.0             82.0           21.92   
3           1496.0             784.0            712.0           73.02   
4            229.0             112.0            117.0           12.97   

   Densidad Poblacion  Altitud Municipios  Nacimientos  Matrimonios  \
0              140.50               568.0         37.0         18.0   
1              103.99               219.0        104.0         37.0   
2                7.66               574.0          1.0          0.0   
3               20.49               333.0         13.0          3.0   
4               17.66               467.0          2.0          0.0   

   Defunciones  Total Regimen SS  Total Contratos  Contratos Hombres  \
0         14.0             784.0              617                453   
1         88.0            4397.0             5861               4843   
2          7.0              30.0               20                 14   
3         10.0             224.0              141                 88   
4          3.0             137.0              125                 90   

   Contratos Mujeres  Total Paro Registrado  Parados Hombres  Parados Mujeres  \
0                164                   1992              935             1057   
1               1018                   8393             3742             4651   
2                  6                     50               16               34   
3                 53                    322              177              145   
4                 35                    139               60               79   

   Nº Establecimientos Agricultura  Nº Establecimientos Industria  \
0                              NaN                            NaN   
1                              NaN                            NaN   
2                              NaN                            NaN   
3                              NaN                            NaN   
4                              NaN                            NaN   

   Nº Establecimientos Construccion  Nº Establecimientos Servicios  \
0                               NaN                            NaN   
1                               NaN                            NaN   
2                               NaN                            NaN   
3                               NaN                            NaN   
4                               NaN                            NaN   

   Nº Centros Sanitarios  Nº Centros Educativos  \
0                    1.0                    4.0   
1                    2.0                    9.0   
2                    1.0                    NaN   
3                    1.0                    2.0   
4                    NaN                    2.0   

   Número de transacciones anuales 2011  Nº Residencias Mayores  \
0                                  25.0                     1.0   
1                                 185.0                     3.0   
2                                   0.0                     NaN   
3                                   8.0                     NaN   
4                                   7.0                     NaN   

   Nº Instituciones Religiosas y Militares  Nº Hombres Extranjeros  \
0                                      1.0                   150.0   
1                                      NaN                   409.0   
2              

In [42]:
Despoblacion.shape

(8116, 78)

In [37]:
Formato = Despoblacion.dtypes
print(Formato.to_string())

Nombre Provincia                            object
Codigo Provincia                            object
Nombre Municipio                            object
Codigo Municipio                            object
Poblacion Total                            float64
Poblacion Hombre                           float64
Poblacion Mujer                            float64
Superficie Km2                             float64
Densidad Poblacion                         float64
Altitud Municipios                         float64
Nacimientos                                float64
Matrimonios                                float64
Defunciones                                float64
Total Regimen SS                           float64
Total Contratos                              int64
Contratos Hombres                            int64
Contratos Mujeres                            int64
Total Paro Registrado                        int64
Parados Hombres                              int64
Parados Mujeres                

Transformamos todos los valores del DataFrame en numéricos con dos decimales.  
Antes de ello, en la fase previa de recopilación de la data, se detectó que había tablas con variables que toman el valor \*.
Estos valores corresponden a variables extraídas del INE donde, por secreto estadístico, los registros que tienen un valor inferior a 5, les asigunan un \*.   
Localizamos las columnas que contienen \* y lo sustituimos por la media:

In [38]:
ColumnasAsterisco = Despoblacion.columns[Despoblacion.isin(['*']).any()]
ColumnasAsterisco

Index(['Edad < 16', '16 =< Edad =<64', 'Edad >= 65', 'Soltero/a', 'Casado/a',
       'Separado/a', 'Divorciado/a', 'Viudo/a', 'Nº de Jubilados',
       'Empresario que emplea personal', 'Empresario que no emplea personal',
       '1 hijo/a', '2 hijos/as', '3 hijos/as', '4 o más hijos/as',
       'Nº de Alquileres', 'Hogares Unipersonales', 'Hogares Familiares',
       '1 persona', '2 personas', '3 personas', '4 personas', '5 personas',
       '6 o más personas', 'Nº Hogares si Internet', 'Nº Hogares no Internet'],
      dtype='object')

In [39]:
Despoblacion[ColumnasAsterisco] = Despoblacion[ColumnasAsterisco].replace('*', 2)

Ahora si que ya podemos realizar el cambio de formato desde la columna número 4 en adelante:

In [40]:
Columnas = Despoblacion.columns[4 : ]
Despoblacion[Columnas] = round(Despoblacion[Columnas].apply(pd.to_numeric), 2)

In [41]:
Formato = Despoblacion.dtypes
print(Formato.to_string())

Nombre Provincia                            object
Codigo Provincia                            object
Nombre Municipio                            object
Codigo Municipio                            object
Poblacion Total                            float64
Poblacion Hombre                           float64
Poblacion Mujer                            float64
Superficie Km2                             float64
Densidad Poblacion                         float64
Altitud Municipios                         float64
Nacimientos                                float64
Matrimonios                                float64
Defunciones                                float64
Total Regimen SS                           float64
Total Contratos                              int64
Contratos Hombres                            int64
Contratos Mujeres                            int64
Total Paro Registrado                        int64
Parados Hombres                              int64
Parados Mujeres                

La dimensión de nuestro DataFrame es de 8.116 filas, coincidiendo con el número de municipios en España en 2011
No obstante, veamos si hay algún Codigo Municipio duplicado:

In [45]:
Despoblacion[Despoblacion['Codigo Municipio'].duplicated()]

Empty DataFrame
Columns: [Nombre Provincia, Codigo Provincia, Nombre Municipio, Codigo Municipio, Poblacion Total, Poblacion Hombre, Poblacion Mujer, Superficie Km2, Densidad Poblacion, Altitud Municipios, Nacimientos, Matrimonios, Defunciones, Total Regimen SS, Total Contratos, Contratos Hombres, Contratos Mujeres, Total Paro Registrado, Parados Hombres, Parados Mujeres, Nº Establecimientos Agricultura, Nº Establecimientos Industria, Nº Establecimientos Construccion, Nº Establecimientos Servicios, Nº Centros Sanitarios, Nº Centros Educativos, Número de transacciones anuales 2011, Nº Residencias Mayores, Nº Instituciones Religiosas y Militares, Nº Hombres Extranjeros, Nº Mujeres Extranjeras, Nº Total Extranjeros, Nº Hombres Otra CCAA, Nº Mujeres Otra CCAA, Nº Total Otra CCAA, Nº edificios, Nº Inmuebles, Nº viviendas, Edad < 16, 16 =< Edad =<64, Edad >= 65, Soltero/a, Casado/a, Separado/a, Divorciado/a, Viudo/a, Sin Estudios, Con Estudios, Nº de Jubilados, Empresario que emplea personal, Empresario que no emplea personal, 1 hijo/a, 2 hijos/as, 3 hijos/as, 4 o más hijos/as, Nº de Alquileres, Hogares Unipersonales, Hogares Familiares, 1 persona, 2 personas, 3 personas, 4 personas, 5 personas, 6 o más personas, Edad Media Edificios, Uso Industrial, Uso Residelcial, Uso Terciario, Otros Usos, Nº Hogares si Internet, Nº Hogares no Internet, Nº Vehiculos, Nº Conductores, Deuda (Miles de €), Total impuestos (Miles de €), Superficie (Ha) Agricolas, Nº Explotaciones Agrícolas, Nº Explotaciones Ganaderas]
Index: []

Eliminamos aquellos municipios que tiene una densidad menor a 12,5, municipios ya despoblados en 2011 dificilmente pueden volver a áreas pobladas:

In [15]:
Despob1acion = Despoblacion.loc[Despoblacion['Densidad Poblacion'] >= 12.5 ]

In [16]:
Despob1acion.head()

Nombre Provincia Codigo Provincia    Nombre Municipio Codigo Municipio  \
0      Araba/Álava               01    Alegría-Dulantzi            01001   
1      Araba/Álava               01             Amurrio            01002   
3      Araba/Álava               01             Aramaio            01003   
4      Araba/Álava               01             Armiñón            01006   
6      Araba/Álava               01  Arrazua-Ubarrundia            01008   

   Poblacion Total  Poblacion Hombre  Poblacion Mujer  Superficie Km2  \
0           2803.0            1466.0           1337.0           19.95   
1          10014.0            5030.0           4984.0           96.30   
3           1496.0             784.0            712.0           73.02   
4            229.0             112.0            117.0           12.97   
6            951.0             508.0            443.0           57.56   

   Densidad Poblacion  Altitud Municipios  Nacimientos  Matrimonios  \
0              140.50               568.0         37.0         18.0   
1              103.99               219.0        104.0         37.0   
3               20.49               333.0         13.0          3.0   
4               17.66               467.0          2.0          0.0   
6               16.52               528.0         11.0          1.0   

   Defunciones  Regimen General  Regimen Agrario  Regimen Mar  Regimen Hogar  \
0         14.0            576.0              5.0          0.0           10.0   
1         88.0           3713.0             15.0          0.0           47.0   
3         10.0            148.0              2.0          0.0            5.0   
4          3.0            115.0              0.0          0.0            0.0   
6          7.0            148.0              2.0          0.0            6.0   

   Regimen Autonomos  Regimen Carbon  Total Regimen SS  Total Contratos  \
0              193.0             0.0             784.0               31   
1              622.0             0.0            4397.0              382   
3               69.0             0.0             224.0                6   
4               22.0             0.0             137.0                3   
6              127.0             0.0             283.0               11   

   Contratos iniciales indefinidos hombres  \
0                                        0   
1                                        3   
3                                        0   
4                                        0   
6                                        0   

   Contratos iniciales temporales hombres  \
0                                      24   
1                                     276   
3                                       4   
4                                       0   
6                                       7   

   Contratos convertidos en indefinidos hombres  \
0                                             0   
1                                            15   
3                                             0   
4                                             0   
6                                             0   

   Contratos iniciales indefinidos mujeres  \
0                                        0   
1                                        1   
3                                        0   
4                                        0   
6                                        0   

   Contratos iniciales temporales mujeres  \
0                                       7   
1                                      82   
3                                       2   
4                                       1   
6                                       4   

   Contratos convertidos en indefinidos mujeres  Contratos  Agricultura  \
0                                             0                       0   
1                                             5                       2   
3                                             0                       0   
4                                           

In [17]:
Despob1acion.shape

(4411, 96)

Analizamos los valores nulos:

In [18]:
Valores = Despob1acion.isnull().sum()
print(Valores.to_string())

Nombre Provincia                                   0
Codigo Provincia                                   0
Nombre Municipio                                   0
Codigo Municipio                                   0
Poblacion Total                                    0
Poblacion Hombre                                   0
Poblacion Mujer                                    0
Superficie Km2                                     0
Densidad Poblacion                                 0
Altitud Municipios                                 1
Nacimientos                                        0
Matrimonios                                        0
Defunciones                                        0
Regimen General                                   17
Regimen Agrario                                   17
Regimen Mar                                       17
Regimen Hogar                                     17
Regimen Autonomos                                 17
Regimen Carbon                                

**Altitud Municipios**: solo un valor nulo.  
Se busca el valor y se completa:

In [19]:
Despob1acion[Despob1acion['Altitud Municipios'].isnull()]

Nombre Provincia Codigo Provincia Nombre Municipio Codigo Municipio  \
2183        Coruña, A               15     Oza dos Ríos            15063   

      Poblacion Total  Poblacion Hombre  Poblacion Mujer  Superficie Km2  \
2183           3229.0            1622.0           1607.0           72.14   

      Densidad Poblacion  Altitud Municipios  Nacimientos  Matrimonios  \
2183               44.76                 NaN         23.0         11.0   

      Defunciones  Regimen General  Regimen Agrario  Regimen Mar  \
2183         45.0            333.0              2.0          2.0   

      Regimen Hogar  Regimen Autonomos  Regimen Carbon  Total Regimen SS  \
2183           14.0              305.0             0.0             656.0   

      Total Contratos  Contratos iniciales indefinidos hombres  \
2183               15                                        1   

      Contratos iniciales temporales hombres  \
2183                                       9   

      Contratos convertidos en indefinidos hombres  \
2183                                             2   

      Contratos iniciales indefinidos mujeres  \
2183                                        0   

      Contratos iniciales temporales mujeres  \
2183                                       2   

      Contratos convertidos en indefinidos mujeres  Contratos  Agricultura  \
2183                                             1                       0   

      Contratos  Industria  Contratos Construcción  Contratos  Servicios  \
2183                     0                       8                     7   

      Total Paro Registrado  Paro hombre edad < 25  Paro hombre edad 25 -45   \
2183                   2512                    183                       585   

      Paro hombre edad >=45  Paro mujer edad < 25  Paro mujer edad 25 -45   \
2183                    427                    42                      733   

      Paro mujer edad >=45  Paro Agricultura  Paro Industria  \
2183                   542                41             376   

      Paro Construcción  Paro Servicios  Paro Sin empleo Anterior  \
2183                422            1361                       312   

      Nº Establecimientos Agricultura  Nº Establecimientos Industria  \
2183                              NaN                            NaN   

      Nº Establecimientos Construccion  Nº Establecimientos Servicios  \
2183                               NaN                            NaN   

      Nº Centros Sanitarios  Nº Centros Educativos  \
2183                    1.0                    NaN   

      Número de transacciones anuales 2011  Nº Residencias Mayores  \
2183                                   NaN                     NaN   

      Nº Instituciones Religiosas y Militares  Nº Hombres Extranjeros  \
2183                                      NaN                    66.0   

      Nº Mujeres Extranjeras  Nº Total Extranjeros  Nº Hombres Otra CCAA  \
2183                    61.0                 127.0                  65.0   

      Nº Mujeres Otra CCAA  Nº Total Otra CCAA  Nº edificios  Nº Inmuebles  \
2183                  52.0               117.0        1857.0        2084.0   

      Nº viviendas  Edad < 16  16 =< Edad =<64  Edad >= 65  Soltero/a  \
2183        2043.0        335             1855        1030       1175   

      Casado/a  Separado/a  Divorciado/a  Viudo/a  Sin Estudios  \
2183      1585           2            80      335          1130   

      Estudios Primarios o Secundarios  Estudios Superiones  Nº de Jubilados  \
2183                              1430                  220             1180   

      Nº de Estudiantes  Empresario que emplea personal  \
2183                300                             195   

      Empresario que no emplea personal  1 hijo/a  2 hijos/as  3 hijos/as  \
2183                                240       305         470         170   

      4 o más hijos/as  Nº de Alquileres  Hogares Unipersonales  \
2183                70                 2      

In [20]:
Despob1acion.loc[2183, 'Altitud Municipios'] = 231

**Afiliados SS**: Los campos Regimen General, Regimen Agrario, Regimen Mar, Regimen Hogar, Regimen Autonomos , Regimen Carbon, Total Regimen SS tienen 17 valores nulos.  
La tabla de afiliados no contiene los 17 municipios correspondientes:

In [21]:
Despob1acion[Despob1acion['Regimen General'].isnull()].head(1)

Nombre Provincia Codigo Provincia Nombre Municipio Codigo Municipio  \
113         Albacete               02      Pozo Cañada            02901   

     Poblacion Total  Poblacion Hombre  Poblacion Mujer  Superficie Km2  \
113           2895.0            1462.0           1433.0          116.83   

     Densidad Poblacion  Altitud Municipios  Nacimientos  Matrimonios  \
113               24.78               830.0         25.0         12.0   

     Defunciones  Regimen General  Regimen Agrario  Regimen Mar  \
113         23.0              NaN              NaN          NaN   

     Regimen Hogar  Regimen Autonomos  Regimen Carbon  Total Regimen SS  \
113            NaN                NaN             NaN               NaN   

     Total Contratos  Contratos iniciales indefinidos hombres  \
113               29                                        2   

     Contratos iniciales temporales hombres  \
113                                      18   

     Contratos convertidos en indefinidos hombres  \
113                                             0   

     Contratos iniciales indefinidos mujeres  \
113                                        0   

     Contratos iniciales temporales mujeres  \
113                                       9   

     Contratos convertidos en indefinidos mujeres  Contratos  Agricultura  \
113                                             0                      12   

     Contratos  Industria  Contratos Construcción  Contratos  Servicios  \
113                     1                       2                    14   

     Total Paro Registrado  Paro hombre edad < 25  Paro hombre edad 25 -45   \
113                   4180                    342                       816   

     Paro hombre edad >=45  Paro mujer edad < 25  Paro mujer edad 25 -45   \
113                    825                   255                     1323   

     Paro mujer edad >=45  Paro Agricultura  Paro Industria  \
113                   619               285             310   

     Paro Construcción  Paro Servicios  Paro Sin empleo Anterior  \
113                917            2297                       371   

     Nº Establecimientos Agricultura  Nº Establecimientos Industria  \
113                              NaN                            NaN   

     Nº Establecimientos Construccion  Nº Establecimientos Servicios  \
113                               NaN                            NaN   

     Nº Centros Sanitarios  Nº Centros Educativos  \
113                    1.0                    NaN   

     Número de transacciones anuales 2011  Nº Residencias Mayores  \
113                                  14.0                     NaN   

     Nº Instituciones Religiosas y Militares  Nº Hombres Extranjeros  \
113                                      NaN                    82.0   

     Nº Mujeres Extranjeras  Nº Total Extranjeros  Nº Hombres Otra CCAA  \
113                    58.0                 140.0                  62.0   

     Nº Mujeres Otra CCAA  Nº Total Otra CCAA  Nº edificios  Nº Inmuebles  \
113                  65.0               127.0         995.0        1355.0   

     Nº viviendas  Edad < 16  16 =< Edad =<64  Edad >= 65  Soltero/a  \
113        1308.0        430             1915         520       1105   

     Casado/a  Separado/a  Divorciado/a  Viudo/a  Sin Estudios  \
113      1560           2             2      140           930   

     Estudios Primarios o Secundarios  Estudios Superiones  Nº de Jubilados  \
113                              1310                  165              510   

     Nº de Estudiantes  Empresario que emplea personal  \
113                480                             105   

     Empresario que no emplea personal  1 hijo/a  2 hijos/as  3 hijos/as  \
113                                 60       105         365         180   

     4 o más hijos/as  Nº de Alquileres  Hogares Unipersonales  \
113               160                15                    140   

     Hogares Familiares  1 persona 

In [22]:
Afiliaciones11 = pd.read_csv('/home/dsc/Python/Despoblacion/Data/06_Output_Afiliados_SS_11.csv' ,dtype='object')

In [23]:
Afiliaciones11[Afiliaciones11['Codigo Municipio'] == '02901']

Empty DataFrame
Columns: [Codigo Municipio, Regimen General, Regimen Agrario, Regimen Mar, Regimen Hogar, Regimen Autonomos, Regimen Carbon, Total Regimen SS]
Index: []

Eliminamos los 17 registros:

In [24]:
Columnas = Afiliaciones11.columns[1:]
Columnas.tolist()

['Regimen General',
 'Regimen Agrario',
 'Regimen Mar',
 'Regimen Hogar',
 'Regimen Autonomos',
 'Regimen Carbon',
 'Total Regimen SS']

In [25]:
Despob1acion = Despob1acion.dropna(subset = Columnas)

**Número de transacciones anuales 2011**: Casualmente son 4 municipios que posteriormente, en 2016,se fusionaron: Cercedo-Cotobade, Oza-Cesuras. Vemos si en el DataFrame de transacciones aparecen los municipios fusionados:


In [26]:
Despob1acion[Despob1acion['Número de transacciones anuales 2011'].isnull()].head()

Nombre Provincia Codigo Provincia Nombre Municipio Codigo Municipio  \
2145        Coruña, A               15          Cesuras            15026   
2183        Coruña, A               15     Oza dos Ríos            15063   
5281       Pontevedra               36          Cerdedo            36011   
5282       Pontevedra               36         Cotobade            36012   

      Poblacion Total  Poblacion Hombre  Poblacion Mujer  Superficie Km2  \
2145           2225.0            1090.0           1135.0           79.46   
2183           3229.0            1622.0           1607.0           72.14   
5281           2297.0            1063.0           1234.0          134.70   
5282           4432.0            2101.0           2331.0           79.84   

      Densidad Poblacion  Altitud Municipios  Nacimientos  Matrimonios  \
2145               28.00               231.0          3.0          9.0   
2183               44.76               231.0         23.0         11.0   
5281               17.05               374.0          7.0          3.0   
5282               55.51               374.0         26.0         12.0   

      Defunciones  Regimen General  Regimen Agrario  Regimen Mar  \
2145         35.0             67.0              6.0          0.0   
2183         45.0            333.0              2.0          2.0   
5281         32.0            119.0              2.0          0.0   
5282         71.0            295.0              9.0          0.0   

      Regimen Hogar  Regimen Autonomos  Regimen Carbon  Total Regimen SS  \
2145           11.0              241.0             0.0             325.0   
2183           14.0              305.0             0.0             656.0   
5281            6.0              107.0             0.0             234.0   
5282           27.0              274.0             0.0             605.0   

      Total Contratos  Contratos iniciales indefinidos hombres  \
2145                3                                        1   
2183               15                                        1   
5281               45                                        0   
5282               12                                        1   

      Contratos iniciales temporales hombres  \
2145                                       0   
2183                                       9   
5281                                      21   
5282                                       4   

      Contratos convertidos en indefinidos hombres  \
2145                                             1   
2183                                             2   
5281                                             0   
5282                                             1   

      Contratos iniciales indefinidos mujeres  \
2145                                        0   
2183                                        0   
5281                                        2   
5282                                        0   

      Contratos iniciales temporales mujeres  \
2145                                       1   
2183                                       2   
5281                                      17   
5282                                       5   

      Contratos convertidos en indefinidos mujeres  Contratos  Agricultura  \
2145                                             0                       0   
2183                                             1                       0   
5281                                             5                       0   
5282                                             1                       0   

      Contratos  Industria  Contratos Construcción  Contratos  Servicios  \
2145                     0                       0                     3   
2183                     0                       8                     7   
5281                    10                       7                    28   
5282                     3                       1                     8   

      Total Paro Registrado  Paro hombre edad < 

In [27]:
Transacciones11 = pd.read_csv('/home/dsc/Python/Despoblacion/Data/12_Output_Transacciones_Inmobiliarias_11.csv' ,dtype='object')

In [28]:
Transacciones11[Transacciones11['Codigo Municipio'] == '36902']#Cercedo-Cotobade

Codigo Municipio Número de transacciones anuales 2011
5344            36902                                  NaN

In [29]:
Transacciones11[Transacciones11['Codigo Municipio'] == '15902'] #Oza-Cesuras

Codigo Municipio Número de transacciones anuales 2011
2218            15902                                 27.0

Eliminamos los registros de Cercedo y Cotobade.   
Tenemos el dato global de Oza-Cesuras, estimamos: la población es un 60/40, asignamos 16/11:

In [30]:
Despob1acion.loc[2145, 'Número de transacciones anuales 2011'] = 11

In [31]:
Despob1acion.loc[2183, 'Número de transacciones anuales 2011'] = 16

In [32]:
Despob1acion = Despob1acion.dropna(subset = ['Número de transacciones anuales 2011'])

**Nº centros sanitarios, educativos, residencias mayores, instituciones religiosas**: Los valores nulos corresponden a la falta de dichos centros en esos municipios.   
Los susbstituimos por ceros:


In [33]:
ColumnasNan = ['Nº Centros Sanitarios', 'Nº Centros Educativos', 
               'Nº Residencias Mayores', 'Nº Instituciones Religiosas y Militares']
Despob1acion[ColumnasNan] = Despob1acion[ColumnasNan].replace(np.nan, 0)

**Nº de extranjeros por municipio**: hay 51 nulos en hombres, 35 en mujeres y 13 en el total.
Además, si nos fijamos en la población total de estos municipios, está muy por debajo de la media.  
Entendemos los valores nulos como cero casos muestrales.

In [34]:
round(Despob1acion['Poblacion Total'].mean(), 2)

10435.39

In [35]:
NulosExtranjeros = Despob1acion[Despob1acion['Nº Total Extranjeros'].isnull()]
(NulosExtranjeros['Poblacion Total'] < 10435.39).sum()

13

In [36]:
NulosExtranjeros['Poblacion Total']

384     145.0
699     298.0
1108    182.0
1567    458.0
1619    482.0
2439    433.0
3626    216.0
4803     55.0
5451    114.0
5539    136.0
7621    227.0
7639    165.0
8028     53.0
Name: Poblacion Total, dtype: float64

In [37]:
ColumnasNan = ['Nº Hombres Extranjeros', 'Nº Mujeres Extranjeras', 'Nº Total Extranjeros']
Despob1acion[ColumnasNan] = Despob1acion[ColumnasNan].replace(np.nan, 0)

**Nº de personas otras comunidades autónomas**: Nº Hombres Otra CCAA hay 8 nulos, Nº Mujeres Otra CCAA hay 4 y en el total no hay nulos.  
Los nulos de Hombres y Mujeres son ceros.

In [38]:
ColumnasNan = ['Nº Hombres Otra CCAA', 'Nº Mujeres Otra CCAA']
Despob1acion[ColumnasNan] = Despob1acion[ColumnasNan].replace(np.nan, 0)

**Edad Media Edificios y número de establecimientos**: Tienen el mismo número de nulos, 409.   
En este caso tenemos muchos registros nulos de los que no disponemos de sus valores.  
Los eliminamos:

In [39]:
Columnas = ['Edad Media Edificios', 'Uso Industrial', 'Uso Residelcial', 'Uso Terciario', 'Otros Usos']
Despob1acion = Despob1acion.dropna(subset = Columnas)

Al eliminar los nulos de edad media de edificios y número de establecimientos, también elimnamos los últimos 4 valores nulos que quedan en el DataFrame correspondiente a las columnas de nº de establecimientos por tipología.  

In [40]:
(Despob1acion.isnull().sum()).sum()

0

In [41]:
Despob1acion.head()

Nombre Provincia Codigo Provincia Nombre Municipio Codigo Municipio  \
51         Albacete               02        Abengibre            02001   
53         Albacete               02         Albacete            02003   
54         Albacete               02         Albatana            02004   
59         Albacete               02          Almansa            02009   
60         Albacete               02           Alpera            02010   

    Poblacion Total  Poblacion Hombre  Poblacion Mujer  Superficie Km2  \
51            902.0             457.0            445.0           30.77   
53         171390.0           84288.0          87102.0         1126.99   
54            794.0             403.0            391.0           30.50   
59          25432.0           12631.0          12801.0          531.82   
60           2438.0            1261.0           1177.0          178.47   

    Densidad Poblacion  Altitud Municipios  Nacimientos  Matrimonios  \
51               29.31               634.0          5.0          3.0   
53              152.08               681.0       1957.0        658.0   
54               26.03               582.0          5.0          3.0   
59               47.82               687.0        254.0        101.0   
60               13.66               852.0         15.0          7.0   

    Defunciones  Regimen General  Regimen Agrario  Regimen Mar  Regimen Hogar  \
51         19.0            149.0             23.0          0.0            2.0   
53       1156.0          55548.0           2403.0          0.0          924.0   
54          5.0             33.0             39.0          0.0            2.0   
59        214.0           5480.0            159.0          0.0           58.0   
60         26.0            210.0             96.0          0.0            5.0   

    Regimen Autonomos  Regimen Carbon  Total Regimen SS  Total Contratos  \
51               92.0             0.0             266.0                9   
53            10621.0             0.0           69496.0             4525   
54               69.0             0.0             143.0               12   
59             1649.0             0.0            7346.0              308   
60              207.0             0.0             518.0               27   

    Contratos iniciales indefinidos hombres  \
51                                        0   
53                                      108   
54                                        0   
59                                        8   
60                                        0   

    Contratos iniciales temporales hombres  \
51                                       7   
53                                    2061   
54                                       9   
59                                     139   
60                                      18   

    Contratos convertidos en indefinidos hombres  \
51                                             0   
53                                           104   
54                                             0   
59                                            20   
60                                             0   

    Contratos iniciales indefinidos mujeres  \
51                                        0   
53                                       73   
54                                        0   
59                                        9   
60                                        0   

    Contratos iniciales temporales mujeres  \
51                                       2   
53                                    2092   
54                                       2   
59                                     115   
60                                       9   

    Contratos convertidos en indefinidos mujeres  Contratos  Agricultura  \
51                                             0                       1   
53                                            87                     472   
54                                             1                       6   

In [42]:
Despob1acion.to_csv('Despoblacion_11.csv', header = True, index = False)